src/Evolution/Systems/Cce/AnalyticSolutions/WorldtubeData.hpp

```diff
@@ -0,0 +1,238 @@
+      double time, ::Tags::dt<gr::Tags::Shift<3, ::Frame::Inertial, DataVector>>
+      /*meta*/) const noexcept;
+
+  virtual void variables_impl(
+      gsl::not_null<tnsr::I<DataVector, 3>*> dr_shift, size_t output_l_max,
+      double time, Tags::Dr<gr::Tags::Shift<3, ::Frame::Inertial, DataVector>>
+      /*meta*/) const noexcept;
+
+  virtual void variables_impl(gsl::not_null<Scalar<DataVector>*> lapse,
+                              size_t output_l_max, double time,
+                              gr::Tags::Lapse<DataVector>
+                              /*meta*/) const noexcept;
+
+  virtual void variables_impl(gsl::not_null<Scalar<DataVector>*> dt_lapse,
+                              size_t output_l_max, double time,
+                              ::Tags::dt<gr::Tags::Lapse<DataVector>>
+                              /*meta*/) const noexcept;
+
+  virtual void variables_impl(gsl::not_null<Scalar<DataVector>*> dr_lapse,
+                              size_t output_l_max, double time,
+                              Tags::Dr<gr::Tags::Lapse<DataVector>>
+                              /*meta*/) const noexcept;
+
+  virtual void variables_impl(
+      gsl::not_null<Scalar<SpinWeighted<ComplexDataVector, -2>>*> news,
+      size_t output_l_max, double time, Tags::News /*meta*/) const noexcept = 0;
+
+  template <typename Tag>
+  struct IntermediateCache {
+    db::item_type<Tag> data;
+    size_t l_max = 0;
+    double time = -std::numeric_limits<double>::infinity();
+  };
+
+  template <typename Tag>
+  struct IntermediateCacheTag {
+    using type = IntermediateCache<Tag>;
+  };
+
+  using IntermediateCacheTuple =
+      tuples::tagged_tuple_from_typelist<tmpl::transform<
+          tmpl::list<
+              Tags::CauchyCartesianCoords,
+              Tags::Dr<Tags::CauchyCartesianCoords>,
+              gr::Tags::SpacetimeMetric<3, ::Frame::Inertial, DataVector>,
+              GeneralizedHarmonic::Tags::Pi<3, ::Frame::Inertial>,
+              GeneralizedHarmonic::Tags::Phi<3, ::Frame::Inertial>,
+              gr::Tags::SpatialMetric<3, ::Frame::Inertial, DataVector>,
+              gr::Tags::Shift<3, ::Frame::Inertial, DataVector>,
+              gr::Tags::Lapse<DataVector>,
+              ::Tags::dt<
+                  gr::Tags::SpacetimeMetric<3, ::Frame::Inertial, DataVector>>,
+              ::Tags::dt<
+                  gr::Tags::SpatialMetric<3, ::Frame::Inertial, DataVector>>,
+              ::Tags::dt<gr::Tags::Shift<3, ::Frame::Inertial, DataVector>>,
+              ::Tags::dt<gr::Tags::Lapse<DataVector>>,
+              Tags::Dr<
+                  gr::Tags::SpatialMetric<3, ::Frame::Inertial, DataVector>>,
+              Tags::Dr<gr::Tags::Shift<3, ::Frame::Inertial, DataVector>>,
+              Tags::Dr<gr::Tags::Lapse<DataVector>>, Tags::News>,
+          tmpl::bind<IntermediateCacheTag, tmpl::_1>>>;
+
+  mutable IntermediateCacheTuple intermediate_cache_;
```

Comment:
Some of the caching could be omitted in favor of recomputing many of the inputs several times depending on the combination requested -- that would make this class comparatively slow, but have similar interface.
I suppose I could have each of the variable generation steps also take an argument where the calling code holds the cache and passes it in to improve the speed, but then the interface becomes more cumbersome by requiring the calling code manage the data storage etc.
This is particularly relevant for a later "analytic" solution that I'll be putting in later that follows this framework -- the Robinson-Trautman solution that actually requires an internal evolved variable to determine the solution. My view is that version really should have internal state, because otherwise the calling code has to perform far too much management of the time-stepped variable.
Can you clarify the question about more than one component? This class is intended to only be held by a different version of the `WorldtubeBoundary` (singleton) component, so the internal state data would not be duplicated across components.